test(quantic): Tabs E2E Cypress to Playwright + unit tests Migration

mmitiche opened the pull request:
[SFINT-5794](https://coveord.atlassian.net/browse/SFINT-5794)

## IN THIS PR:

### Quantic Tab Bar component:
- Added a `tabBarItem` slot name for the quantic Tab slot
- Added unit tests for the `quanticTabBar` component
- Added Playwright E2E tests for the `quanticTabBar` component

### Quantic Tab component:
- Added unit tests for the `quanticTab` component
- Added Playwright E2E tests for the `quanticTab` component

### UNIT TESTS:

QUANTIC TAB:

<img width="720" alt="image" src="https://github.com/user-attachments/assets/be94f71b-1910-495d-a340-83532adc702a" />

QUANTIC TAB BAR:

<img width="720" alt="image" src="https://github.com/user-attachments/assets/44d9dff0-215b-40c0-b1d9-3fe4142a2b60" />


### E2E PLAYWRIGHT TESTS:

QUANTIC TAB:

<img width="1020" alt="image" src="https://github.com/user-attachments/assets/2dbce88c-d25d-458f-acab-1bf968977010" />


QUANTIC TAB BAR:

<img width="1020" alt="image" src="https://github.com/user-attachments/assets/dc004d5b-74e9-48d7-b876-e8025a0aeb6e" />

[SFINT-5794]: https://coveord.atlassian.net/browse/SFINT-5794?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

erocheleau commented:
@mmitiche do you mind as part of one of the two PRs fix the readme please?

The readme says this:

```
To run Playwright tests, run:

npm run e2e:playwright:
```

There is a `:` at the end that shouldn't be there I think :) I may have missed it in the previous PR.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/__tests__/quanticTabBar.test.js:
```diff
@@ -0,0 +1,123 @@
+    expect(dropdown.classList).not.toContain(expectedOpenDropdownClass);
+  });
+
+  describe('when the light theme property is set to true', () => {
+    it('should display the component with the light theme styles', async () => {
+      const expectedDarkThemeClass = 'slds-theme_shade';
+      const element = createTestComponent({lightTheme: true});
+      await flushPromises();
+
+      const tabBarContainer = element.shadowRoot.querySelector(
+        selectors.tabBarContainer
+      );
+      expect(tabBarContainer.classList).not.toContain(expectedDarkThemeClass);
+    });
+  });
+
+  describe('when the light theme property is set to false', () => {
+    it('should display the component with the light theme styles', async () => {
+      const expectedDarkThemeClass = 'slds-theme_shade';
+      const element = createTestComponent({lightTheme: false});
+      await flushPromises();
+
+      const tabBarContainer = element.shadowRoot.querySelector(
+        selectors.tabBarContainer
+      );
+      expect(tabBarContainer.classList).toContain(expectedDarkThemeClass);
+    });
+  });
```
I don't know if you agree but I find we always have too many `describe` xD

```suggestion
  describe('the lightTheme property', () => {
    it('when true, should display the component as light theme', async () => {
      const expectedDarkThemeClass = 'slds-theme_shade';
      const element = createTestComponent({lightTheme: true});
      await flushPromises();

      const tabBarContainer = element.shadowRoot.querySelector(
        selectors.tabBarContainer
      );
      expect(tabBarContainer.classList).not.toContain(expectedDarkThemeClass);
    });

    it('when false, should display the component as dark theme', async () => {
      const expectedDarkThemeClass = 'slds-theme_shade';
      const element = createTestComponent({lightTheme: false});
      await flushPromises();

      const tabBarContainer = element.shadowRoot.querySelector(
        selectors.tabBarContainer
      );
      expect(tabBarContainer.classList).toContain(expectedDarkThemeClass);
    });
  });
```

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts:
```diff
@@ -0,0 +1,74 @@
+  }
+
+  get allVisibleTabs(): Locator {
+    return this.tabBar.locator('button.slds-tabs_default__item').filter({
```
`getByRole` ?

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:
```diff
@@ -0,0 +1,199 @@
+  {
+    label: 'extra small',
+    width: extraSmallViewportWidth,
+    primaryExpectation: 'the selected tab ',
+    secondaryExpectation: 'and the more button should be shrunk',
```
Superfluous to create two variables there

```suggestion
    expectationDescription: 'should display only the selected tab, the other tabs should be displayed in the dropdown list and the more button should be shrunk'
```

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:
```diff
@@ -0,0 +1,199 @@
+    width: extraSmallViewportWidth,
+    primaryExpectation: 'the selected tab ',
+    secondaryExpectation: 'and the more button should be shrunk',
+    numberOfTabsDisplayed: 1,
+    numberOfOptionsInDropdown: 3,
+    moreButtonLabel: '',
```
```suggestion
    expectedNumberOfTabsDisplayed: 1,
    expectedNumberOfOptionsInDropdown: 3,
    expectedMoreButtonLabel: '',
```

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts:
```diff
@@ -0,0 +1,74 @@
+    return this.moreButton.locator('button').first();
+  }
+
+  get allDropdownOptions(): Locator {
```
```suggestion
  get allTabsInDropdown(): Locator {
```

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:
```diff
@@ -0,0 +1,199 @@
+
+            await tabBar.clickMoreButton();
+
+            const dropdownOptions = tabBar.allDropdownOptions;
+            expect(dropdownOptions).not.toBeNull();
+            expect(await dropdownOptions.count()).toEqual(
+              numberOfOptionsInDropdown
+            );
```
"Options" is fairly generic, I think it's clearer to specify these are tabs.

```suggestion
            const tabsInDropdown = tabBar.allTabsInDropdown;
            expect(tabsInDropdown).not.toBeNull();
            expect(await tabsInDropdown.count()).toEqual(
              numberOfTabsInDropdown
            );
```

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:
```diff
@@ -0,0 +1,199 @@
+      test.use({
+        viewport: {width: mediumViewportWidth, height: standardViewportHeight},
+      });
+      test('should open the dropdown and display the options', async ({
+        tabBar,
+      }) => {
+        const expectedNumberOfDropdownOptions = 2;
+        await tabBar.clickMoreButton();
+
+        const dropdownOptions = tabBar.allDropdownOptions;
+        expect(dropdownOptions).not.toBeNull();
+        expect(await dropdownOptions.count()).toEqual(
+          expectedNumberOfDropdownOptions
+        );
+      });
```
You already tested that no? You already tested a medium width in the previous loop

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:
```diff
@@ -0,0 +1,199 @@
+      });
+    });
+
+    test.describe('when the dropdown loses focus', () => {
```
That really looks like what should be a unit test and not a e2e.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:
```diff
@@ -0,0 +1,199 @@
+      });
+
+      test.describe('when a tab is selected from the dropdown list', () => {
+        test('should correctly select the tab and make it active', async ({
```
That's really close to a unit test if it isn't one.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:
```diff
@@ -0,0 +1,199 @@
+      test.use({
+        viewport: {width: mediumViewportWidth, height: standardViewportHeight},
+      });
+      test('should open the dropdown and display the options', async ({
```
This also looks like it should be a unit test.

erocheleau reviewed (changes requested):
Some suggested improvements

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/__tests__/quanticTabBar.test.js:
```diff
@@ -0,0 +1,123 @@
+    );
+    await flushPromises();
+
+    exampleSlots.forEach((tabItem, index) => {
+      expect(tabItem.label).toEqual(exampleSlots[index].label);
+      expect(tabItem.engineId).toEqual(exampleSlots[index].engineId);
+      expect(tabItem.expression).toEqual(exampleSlots[index].expression);
+      expect(tabItem.isActive).toEqual(exampleSlots[index].isActive);
+    });
```
what's happening here is checking that `exampleSlots` is equal to `exampleSlots` 🤨

mmitiche commented:
I added more unit tests for the Quantic Tab Bar component, this component in a bit complicated component, so we should use as much unit tests as possible to test it, we should invest more effort in unit tests in PRs like this.

I have added to this PR the logic to test how the component behaves, by mocking and playing with the width of the container and the slots passed to this component.

mmitiche commented on packages/quantic/decisions/0001-testing-strategy.md:
```diff
@@ -39,12 +39,15 @@ Key points for Playwright E2E tests:
 - A user navigates through a whole search workflow, interacting with multiple components. We test the expected Search API calls and responses, the analytics calls and response, and also the reactivity of the full search experience.
 - Verify the integration between the Quantic components and Salesforce. We test that modifying data through the Salesforce components has an impact and triggers the correct components reactions in a Quantic experience.
+- A user modifies the URL parameters. We assess the impact on our components.
+- The browser viewport size changes. We verify that our components adapt their display accordingly.
```
This point is to be clarified as you can see in this PR, the viewport size logic can be tested with unit tests and sometimes is better tested in unit tests, I see the differentiation as the following: 
- The logic based on the viewport size that makes the component render its element can be be unit tested and is sometimes better unit tested like in this PR.
- The reactivity of the component to the resize events of the browser is something that can be E2E and is better E2E tested.

Bottom line is saying that only E2E tests are used to test such logics is untrue, as I'm noticing while evaluating how the QuanticTabBar component, the unit tests in this PR seem more useful.

erocheleau replied:
The VIEWPORT cannot be tested in unit tests, but the size of the container of an element can definitely be tested via unit tests.

That's why I suggested that since this was a bit more related to reactivity (aka how many tabs can be shown in a specific width) we should do those via playwright tests.

But I agree it's in a gray zone.

mmitiche commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/fixture.ts:
```diff
@@ -0,0 +1,63 @@
+};
+
+type QuanticTabBarE2ESearchFixtures = QuanticTabBarE2EFixtures & {
+  urlHash: string;
```
The tab bar should  not need a urlHash

mmitiche commented:
@SimonMilord  I added unit tests for the QuanticTab component and the QuanticTabBar component, I will let @erocheleau review the rest!

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/quanticTabBar.js:
```diff
@@ -357,7 +357,10 @@ export default class QuanticTabBar extends LightningElement {
+    /** @type {HTMLSlotElement} */
+    const slot = this.template.querySelector('slot');
+
+    return Array.from(Array.from(slot.assignedElements())).filter((element) =>
```
Why do you even need an Array.from()? `slot.assignedElements()` already return an Element[] it looks like?

SimonMilord replied:
you are right, not sure why it was added, but it works without so Ill remove it

erocheleau reviewed (commented):
Only 2 or 3 questions left I think

erocheleau commented on packages/quantic/cypress/e2e/default-1/tab-bar/tab-bar-selectors.ts:
```diff
@@ -21,7 +21,8 @@ export const TabBarSelectors: TabBarSelector = {
   activeTab: () =>
     TabBarSelectors.get().find('button.slds-tabs_default__item.slds-is-active'),
-  moreButton: () => TabBarSelectors.get().find('.tab-bar_more-button'),
+  moreButton: () =>
```
So that one didn't work with the `getByRole` I assume?

SimonMilord replied:
nope, thats why I settled for the `getByTestId`

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/__tests__/quanticTabBar.test.js:
```diff
@@ -0,0 +1,337 @@
+      );
+    });
+
+    describe('reactivity to window resize', () => {
```
You already have this describe on line 143, I think these tests should be moved in the same describe group.

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/pageObject.ts:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  get moreButton(): Locator {
+    return this.tabBar.getByTestId('tab-bar_more-button');
```
That didn't work with the `getByRole` right? I think I remember trying to make it work...

SimonMilord replied:
yeah I tried it but it worked with `getByTestId`

erocheleau commented on packages/quantic/force-app/main/default/lwc/quanticTabBar/e2e/quanticTabBar.e2e.ts:
```diff
@@ -0,0 +1,178 @@
+      });
+
+      test.describe('when a tab is selected from the dropdown list', () => {
+        test('should correctly select the tab and make it active', async ({
+          tabBar,
+        }) => {
+          const expectedSelectedTabLabel = 'Tab 4';
+          const expectedNumberOfDropdownOptions = 2;
+          await tabBar.clickMoreButton();
+
+          const dropdownOptionsCount = await tabBar.allDropdownOptions.count();
+          expect(dropdownOptionsCount).not.toBeNull();
+          expect(dropdownOptionsCount).toEqual(expectedNumberOfDropdownOptions);
+
+          await tabBar.clickDropdownOption(1);
+
+          const activeTab = await tabBar.activeTab.textContent();
+
+          expect(activeTab).not.toBeNull();
+          expect(activeTab).toEqual(expectedSelectedTabLabel);
+        });
+
+        test('should trigger a new search send the correct UA analytics event', async ({
```
Merge these two tests:

`test('should make the tab active, trigger a new search and send the correct analytics event')`

You're performing the same "action" in order to test something, so test everything that should happen from the same action.

erocheleau reviewed (approved):
Some small comment and questions, but almost ready to merge!
